PSWebServiceLibrary.php

```diff
@@ -92,50 +90,66 @@ function __construct($url, $key, $debug = true)
+            case 100:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive100ContinueStatus();
+            case 101:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive101SwitchingProtocolsStatus();
+            case 102:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive102ProcessingStatus();
+            case 103:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive103EarlyHintsStatus();
             case 204:
-                $error_message = 'No content';
-                break;
+                throw new PrestaShopWebserviceNoContentException();
+            case 300:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive300MultipleChoicesStatus();
+            case 301:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive301MovedPermanentlyStatus();
+            case 302:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive302FoundStatus();
+            case 303:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive303SeeOtherStatus();
+            case 304:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive304NotModifiedStatus();
+            case 307:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive307TemporaryRedirectStatus();
+            case 308:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive308PermanentRedirectStatus();
             case 400:
-                $error_message = 'Bad Request';
-                break;
+                throw new PrestaShopWebserviceBadRequestException();
             case 401:
-                $error_message = 'Unauthorized';
-                break;
+                throw new PrestaShopWebserviceUnauthorizedException();
+            case 403:
+                throw new PrestaShopWebserviceForbiddenException();
             case 404:
-                $error_message = 'Not Found';
-                break;
+                throw new PrestaShopWebserviceNotFoundException();
             case 405:
-                $error_message = 'Method Not Allowed';
-                break;
-            case 500:
-                $error_message = 'Internal Server Error';
-                break;
-            default:
-                throw new PrestaShopWebserviceException(
-                    'This call to PrestaShop Web Services returned an unexpected HTTP status of:' . $request['status_code']
-                );
+                throw new PrestaShopWebserviceMethodNotAllowedException();
+            case 429:
+                throw new PrestaShopWebserviceTooManyRequestsException();
+        }
+        if ($statusCode >= 100 && $statusCode < 200
+            || $statusCode >= 202 && $statusCode < 300
+            || $statusCode >= 300 && $statusCode < 500
+            || $statusCode >= 600
+            || $statusCode < 100
+        ) {
```

Comment:
There could be a lot of other things between the API client and the Prestashop API, like reverse proxies, firewalls, HTTP caches.

Those conditions list are here to provide a predictable behavior by the client, whichever infrastructure coule be deployed.

Reply:
Yeah but I'm not talking about the error message, but the condition itself. If you look closely at this if and compare it to the switch above and the other if below, the only possible value for $statusCode is 200 and 201 (which I missed before).

By reordering the conditions, the content of this function should look like this:
```php
if (in_array($statusCode, array(200, 201)) {
    // success, early return
    return;
}
switch ($statusCode) {
    // specific errors as you wrote them
    // [...]
}
// error status
if ($statusCode >= 500 && $statusCode < 600) {
    throw new PrestaShopWebserviceServerException('This call to PrestaShop Web Services responded with a status code indicating it is not available or under a too heavy load to process your request.', $statusCode);
}
// by now all other options have been exhausted, so this is a generic error
throw new PrestaShopWebserviceClientException('This call to PrestaShop Web Services responded with a non-standard code or a code this client could not handle properly. This can come from your web server or reverse proxy configuration.', $statusCode);

Reply:
I'll change the code with an early return 👍